tests/test-snapd-desktop-integration.c

```diff
@@ -0,0 +1,352 @@
+    g_autofree gchar *json = get_json (message->request_body);
+    g_assert_cmpstr(json, ==, "{\"gtk-themes\":[\"GtkTheme2\"],\"icon-themes\":[],\"sound-themes\":[\"SoundTheme2\"]}");
+    send_snapd_response(message, 200, "{\"type\":\"async\", \"change\": \"1234\"}");
+    state = STATE_NOTIFY_COMPLETE;
+    break;
+  default:
+    break;
+  }
+}
+
+static void
+handle_snapd_get_changes_request(SoupMessage *message)
+{
+  send_snapd_response(message, 200, "{\"type\":\"sync\",\"status-code\":200,\"status\":\"OK\",\"result\":{\"id\":\"1234\",\"ready\":true}}");
+}
+
+static void
+handle_snapd_request(SoupServer *server, SoupMessage *message, const char *path, GHashTable *query, SoupClientContext *client, gpointer user_data)
+{
+   if (strcmp(path, "/v2/accessories/themes") == 0) {
+     handle_snapd_themes_request(message);
+   } else if (strcmp(path, "/v2/changes/1234") == 0 && strcmp(message->method, "GET") == 0) {
+     handle_snapd_get_changes_request(message);
+   } else {
+     send_snapd_response(message, 404, "{\"type\":\"error\",\"status-code\":404,\"status\":\"Not Found\",\"result\":{\"message\":\"not found\"}}");
+   }
+}
+
+static gboolean
+setup_mock_snapd(GError **error)
+{
+   snapd_socket_path = g_build_filename(temp_dir, "snapd.socket", NULL);
+
+   g_autoptr(GSocket) socket = g_socket_new(G_SOCKET_FAMILY_UNIX, G_SOCKET_TYPE_STREAM, G_SOCKET_PROTOCOL_DEFAULT, error);
+   if (socket == NULL) {
+     g_prefix_error(error, "Failed to make socket: ");
+     return FALSE;
+   }
+   g_autoptr(GSocketAddress) address = g_unix_socket_address_new(snapd_socket_path);
+   if (!g_socket_bind(socket, address, TRUE, error)) {
+     g_prefix_error(error, "Failed to make socket: ");
+     return FALSE;
+   }
+   if (!g_socket_listen(socket, error)) {
+     g_prefix_error(error, "Failed to listen: ");
+     return FALSE;
+   }
+
+   snapd_server = soup_server_new(SOUP_SERVER_SERVER_HEADER, "MockSnapd/1.0", NULL);
+   soup_server_add_handler(snapd_server, NULL, handle_snapd_request, NULL, NULL);
+   if (!soup_server_listen_socket(snapd_server, socket, 0, error)) {
+     g_prefix_error(error, "Failed to listen for HTTP requests: ");
+     return FALSE;
+   }
+
+   return TRUE;
+}
+
+static gboolean setup_session_bus(GError **error)
+{
+   int address_pipe_fds[2];
+   if (!g_unix_open_pipe(address_pipe_fds, FD_CLOEXEC, error)) {
+      g_prefix_error(error, "Failed to open pipe for D-Bus bus: ");
+      return FALSE;
+   }
+   g_autoptr(GSubprocessLauncher) launcher = g_subprocess_launcher_new(G_SUBPROCESS_FLAGS_STDOUT_SILENCE | G_SUBPROCESS_FLAGS_STDERR_SILENCE);
+   g_subprocess_launcher_take_fd(launcher, address_pipe_fds[1], address_pipe_fds[1]);
+   g_autofree gchar *address_fd_arg = g_strdup_printf("%d", address_pipe_fds[1]);
+   dbus_subprocess = g_subprocess_launcher_spawn(launcher, error, "dbus-daemon", "--nofork", "--session", "--print-address", address_fd_arg, NULL);
```

Comment:
Is there any particular benefit to doing this setup manually rather than using GTestDBus from libgio?

Reply:
There was no benefit, I just wasn't aware of GTestDBus at the time. I've tried replacing this with GTestDBus, but I'm not having much success. I'll land the changes as is as it works fine as is, with just a bit of (hopefully) avoidable complexity. We can replace this code in the future if necessary.